### PR TITLE
Fix Task manager race conditions issues

### PR DESCRIPTION
**Pull request by troopa81**
## Description

This PR fixes differrent race condition issues is QgsTaskManager:

c380b09945  When a task is completed, statusChanged signal is emitted in task thread. As a consequence cleanupAndDeleteTask will be called in the main thread and the task will be deleted (later). It happens sometimes that instructions after statusChanged are executed after the task had been effectively deleted, leading to crash or non emitted signals.

It cause also issue with sub task. Indeed, when a subTask is completed, you have no garantie that instructions after statusChanged are executed before the parent caught the statusChanged signal and delete itself and the subtask.

This commit propose to force  `processSubTasksForCompletion` and `processSubTasksForTermination` execution in main thread (`processSubTasksForHold` if out of the scope because it doesn't delete tasks and is executed in main thread), so all status modification and deletion are made in a determinist order. 

a79431d5ad : remove unlock from `terminated`/`completed` because unlock call from `terminated` crashes the application when called from `cancel` method on not started task. In this case, the task is canceled without being started, so the mutex isn't locked. And the Qt documentation says

> Unlocking a mutex that is not locked results in undefined behavior.

5e4c1ea6a5 : cancel is called from a different thread than the one running the task so mShouldTerminate should be protected with a mutex

956b142239 : we need to be sure that the task is really completly finished (start method has exited) before deleting the task object to avoid crash

0ed9ea93f5 : I fix the tests, slip subTask in several one to ease debug (when it fails on CI for instance) and remove the SKIP tests on CI. I run all tests on linux and Windows, [repeatdly](https://github.com/troopa81/QGIS/commit/3a2382792c) and injecting latency in task manager to try to make it crash. So far, I have no more failing tests.

### CODE REVIEW

File: src/core/qgstaskmanager.cpp [hunk 1]
```diff
@@ -254,7 +264,7 @@ void QgsTask::setProgress( double progress )
 void QgsTask::completed()
 {
   mStatus = Complete;
-  processSubTasksForCompletion();
+  QMetaObject::invokeMethod( this, &QgsTask::processSubTasksForCompletion, Qt::AutoConnection );
```
Comment (elpaso) [inline src/core/qgstaskmanager.cpp]: If I'm not wrong, AutoConnection is the default anyway.

**Comment by elpaso** [comment]
Nice work!

File: src/core/qgstaskmanager.cpp (hunk 1 above)
Comment (troopa81) [reply src/core/qgstaskmanager.cpp]: Right! and the call with a method pointer doesn't exist yet in Qt 5.9 (failing the CI). I changed it.

**Comment by troopa81** [comment]
> Nice work!

Thanks

**Comment by rouault** [comment]
@troopa81 I think for automatic backport to work, I need to squash when merging ? Is it OK for you ?

**Comment by m-kuhn** [comment]
@rouault it should also work with merge commits meanwhile. (https://github.com/Gaurav0/backport/pull/41)

**Comment by rouault** [comment]
OK, I took the "Merge commit" road

**Comment by 3nids** [comment]
sadly, the bot was broken and is back to former version which doesn't work for merge commits.

**Comment by rouault** [comment]
@troopa81 Please issue a manual PR for the 3.12 backport

**Comment by nyalldawson** [comment]
@troopa81 there's a test intermittently failing here still: see https://travis-ci.org/github/qgis/QGIS/jobs/662189227?utm_medium=notification&utm_source=github_status

**Comment by troopa81** [comment]
@nyalldawson  I'll take a look

**Comment by troopa81** [comment]
@nyalldawson fixed #35108

**Pull request merged**